module4/lessons/Week2/BuildAnAPI.md

```diff
@@ -0,0 +1,403 @@
+
+</section>
+
+<section class='answer' markdown='1'>
+
+### 2. Implement the Endpoint
+
+```c#
+[HttpPost]
+public ActionResult CreateBook(Book book)
+{
+    if (!ModelState.IsValid)
+    {
+        Response.StatusCode = 400;
+        return;
+    }
+    _context.Books.Add(book);
+    _context.SaveChanges();
+
+    var savedBook = _context.Books.Last();
+
+    Response.StatusCode = 201;
+    return new JsonResult(savedBook);
+}
+```
+
+</section>
+
+<section class='answer' markdown='1'>
+
+### 3. Experiment with Postman
+
+Let's experiment with this new functionality in Postman.
+
+1. Open postman and create a post request
+2. Add a JSON body to the request
+3. Send the Request - you should receive a 201 response status!
+![](/assets/images/module4/week2/postman_postbooks.png)
+4. Verify the post by running a GetBooks request
+![](/assets/images/module4/week2/postman_getnewbooks.png)
+
+</section>
+
+### PUT /books/1
+
+<section class='call-to-action' markdown='1'>
+**With a partner: Implement an API Endpoint that satisfies these specifications**  
+
+<section class='note' markdown='1'>
+* Path/Method: PUT /api/books/1
+* Status Code: 204
+* Request Content: JSON with the ID of a book to edit, and the field(s) to be edited.
+</section>
+
+Use any resources you can think of to help you along the way: google, .NET documentation, and prior projects are all good resources!
+
+Testing for database updates is tough - try your best, and take a look one solution below, if you are stuck.
+
+<section class='answer' markdown='1'>
+
+### One way to test Update
+```c#
+[Fact]
+public async void PutBook_UpdatesDatabaseRecord()
+{
+    Book book1 = new Book { Title = "Dune", Description = "Sand Monsters" };
+
+    GoodBooksApiContext context = GetDbContext();
+    context.Books.Add(book1);
+    context.SaveChanges();
+
+    HttpClient client = _factory.CreateClient();
+    var jsonString = "{ \"Id\":\"1\", \"Title\":\"Beetlejuice\", \"Description\":\"Sand Monsters\" }";
+    var requestContent = new StringContent(jsonString, Encoding.UTF8, "application/json");
+    var response = await client.PutAsync("/api/books/1", requestContent);
+
+    // Clear all previously tracked DB objects to get a new copy of the updated book
+    context.ChangeTracker.Clear();
+
+    Assert.Equal(204, (int)response.StatusCode);
+    Assert.Equal("Beetlejuice", context.Books.Find(1).Title);
+}
+```
+
+</section>
+
+<section class='note' markdown='1'>
+* 🌶️ If the action creates a NEW book, return a status code 201.
+</section>
+
+</section>
+
+
+### DELETE /books/1
+
+<section class='call-to-action' markdown='1'>
+
+**Individually: Implement an API Endpoint that satisfies these specifications**  
+<section class='note' markdown='1'>
+* Path/Method: DELETE /api/books/1
+* Status Code: 204
+</section>
```

Comment:
The formatting appears off in this section for some reason. The bullets do not render correctly.